File: src/accessiweather/provider_normalization.py
"""Provider-neutral normalization helpers for weather API adapters."""

from __future__ import annotations

from dataclasses import dataclass

from .utils.temperature_utils import TemperatureUnit, calculate_dewpoint
from .weather_client_parsers import (
    convert_f_to_c,
    convert_wind_speed_to_mph_and_kph,
    normalize_pressure,
    normalize_temperature,
)

KM_PER_MILE = 1.609344
MB_PER_INHG = 33.8639


@dataclass(frozen=True)
class TemperaturePair:
    """A temperature represented in both AccessiWeather display units."""

    fahrenheit: float | None
    celsius: float | None


@dataclass(frozen=True)
class SpeedPair:
    """A speed represented in both AccessiWeather display units."""

    mph: float | None
    kph: float | None


@dataclass(frozen=True)
class PressurePair:
    """A pressure represented in both AccessiWeather display units."""

    inches: float | None
    millibars: float | None


@dataclass(frozen=True)
class VisibilityPair:
    """A visibility distance represented in both AccessiWeather display units."""

    miles: float | None
    kilometers: float | None


@dataclass(frozen=True)
class ApparentTemperatureClassification:
    """Feels-like temperature split into chill/index fields."""

    wind_chill_f: float | None = None
    wind_chill_c: float | None = None
    heat_index_f: float | None = None
    heat_index_c: float | None = None


def as_float(value: object) -> float | None:
    """Return ``value`` as a float, treating missing or invalid values as absent."""
    if value is None or value == "":
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def normalize_humidity_percent(value: object, *, fraction: bool = False) -> int | None:
    """Normalize humidity to a rounded 0-100 percentage."""
    numeric = as_float(value)
    if numeric is None:
        return None
    if fraction:
        numeric *= 100
    return round(numeric)


def normalize_temperature_pair(value: object, unit: str | None) -> TemperaturePair:
    """Normalize a provider temperature value to Fahrenheit and Celsius."""
    numeric = as_float(value)
    if numeric is None:
        return TemperaturePair(None, None)
    temp_f, temp_c = normalize_temperature(numeric, unit)
    return TemperaturePair(temp_f, temp_c)


def normalize_dewpoint_pair(
    dewpoint_value: object,
    dewpoint_unit: str | None,
    *,
    fallback_temperature_f: float | None,
    humidity_percent: int | None,
) -> TemperaturePair:
    """Normalize dewpoint, calculating it from temperature and humidity when absent."""
    dewpoint = normalize_temperature_pair(dewpoint_value, dewpoint_unit)
    if dewpoint.fahrenheit is not None or dewpoint.celsius is not None:
        return dewpoint

    if fallback_temperature_f is None or humidity_percent is None:
        return TemperaturePair(None, None)

    dewpoint_f = calculate_dewpoint(
        fallback_temperature_f,
        humidity_percent,
        unit=TemperatureUnit.FAHRENHEIT,
    )
    return TemperaturePair(dewpoint_f, convert_f_to_c(dewpoint_f))


def normalize_pressure_pair(value: object, unit: str | None) -> PressurePair:
    """Normalize pressure to inches of mercury and millibars."""
    numeric = as_float(value)
    if numeric is None:
        return PressurePair(None, None)
    pressure_in, pressure_mb = normalize_pressure(numeric, unit)
    return PressurePair(pressure_in, pressure_mb)


def normalize_pressure_to_pascals(value: object, unit: str | None) -> float | None:
    """Normalize a pressure value to Pascals for NWS-compatible mapped payloads."""
    numeric = as_float(value)
    if numeric is None:
        return None

    unit_text = (unit or "").strip().lower()
    if "hpa" in unit_text or "mb" in unit_text:
        return numeric * 100
    if "pa" in unit_text:
        return numeric
    if "inch" in unit_text or unit_text in {"in", "inhg"}:
        return numeric * 3386.39
    return numeric * 100


def normalize_millibars(value: object) -> PressurePair:
    """Normalize provider values that are already millibars/hectopascals."""
    numeric = as_float(value)
    if numeric is None:
        return PressurePair(None, None)
    return PressurePair(numeric / MB_PER_INHG, numeric)


def _canonical_wind_unit(unit: str | None) -> str | None:
    unit_text = (unit or "").strip().lower()
    if unit_text in {"m/s", "meter/s", "meters/s", "metre/s", "metres/s"}:
        return "wmoUnit:m_s-1"
    if unit_text in {"km/h", "kmh", "kph"}:
        return "wmoUnit:km_h-1"
    if unit_text in {"mph", "mi/h"}:
        return "wmoUnit:mi_h-1"
    if unit_text in {"kn", "kt", "knot", "knots"}:
        return "wmoUnit:kn"
    return unit


def normalize_speed_pair(value: object, unit: str | None) -> SpeedPair:
    """Normalize wind speed to miles per hour and kilometers per hour."""
    numeric = as_float(value)
    if numeric is None:
        return SpeedPair(None, None)
    mph, kph = convert_wind_speed_to_mph_and_kph(numeric, _canonical_wind_unit(unit))
    return SpeedPair(mph, kph)


def format_speed(value: object, unit_label: str) -> str | None:
    """Format provider-native speed for text forecast fields."""
    numeric = as_float(value)
    if numeric is None:
        return None
    return f"{round(numeric)} {unit_label}"


def normalize_visibility_pair(
    value: object,
    unit: str | None,
    *,
    cap_miles: float | None = None,
) -> VisibilityPair:
    """Normalize visibility to miles and kilometers."""
    numeric = as_float(value)
    if numeric is None:
        return VisibilityPair(None, None)

    unit_text = (unit or "").strip().lower().replace(" ", "_")
    if "ft" in unit_text or "feet" in unit_text:
        miles = numeric / 5280
    elif "km" in unit_text or "kilometer" in unit_text or "kilometre" in unit_text:
        miles = numeric / KM_PER_MILE
    elif unit_text in {"mi", "mile", "miles"}:
        miles = numeric
    else:
        miles = numeric / 1609.344

    if cap_miles is not None:
        miles = min(miles, cap_miles)
    return VisibilityPair(miles, miles * KM_PER_MILE)


def classify_apparent_temperature(
    temperature_f: float | None,
    apparent_f: float | None,
    apparent_c: float | None,
) -> ApparentTemperatureClassification:
    """Classify a feels-like value as wind chill or heat index when it differs."""
    if apparent_f is None or temperature_f is None:
        return ApparentTemperatureClassification()
    if apparent_f < temperature_f:
        return ApparentTemperatureClassification(wind_chill_f=apparent_f, wind_chill_c=apparent_c)
    if apparent_f > temperature_f:
        return ApparentTemperatureClassification(heat_index_f=apparent_f, heat_index_c=apparent_c)
    return ApparentTemperatureClassification()


def pirate_temperature_unit(units: str) -> str:
    """Return the temperature unit used by a Pirate Weather unit group."""
    return "F" if units == "us" else "C"


def pirate_wind_unit(units: str) -> str:
    """Return the wind-speed unit used by a Pirate Weather unit group."""
    if units in {"us", "uk2"}:
        return "mph"
    if units == "ca":
        return "km/h"
    return "m/s"


def pirate_visibility_unit(units: str) -> str:
    """Return the visibility unit used by a Pirate Weather unit group."""
    return "mi" if units in {"us", "uk2"} else "km"

File: rust/crates/aw-core/src/provider_normalization.rs
//! Provider-neutral normalization helpers, ported from
//! `accessiweather.provider_normalization` (plus `calculate_dewpoint` from
//! `accessiweather.utils.temperature_utils`).

use crate::py;
use crate::weather_client_parsers::{
    convert_f_to_c, convert_wind_speed_to_mph_and_kph, normalize_pressure, normalize_temperature,
};

pub const KM_PER_MILE: f64 = 1.609344;
pub const MB_PER_INHG: f64 = 33.8639;

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct TemperaturePair {
    pub fahrenheit: Option<f64>,
    pub celsius: Option<f64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct SpeedPair {
    pub mph: Option<f64>,
    pub kph: Option<f64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct PressurePair {
    pub inches: Option<f64>,
    pub millibars: Option<f64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct VisibilityPair {
    pub miles: Option<f64>,
    pub kilometers: Option<f64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct ApparentTemperatureClassification {
    pub wind_chill_f: Option<f64>,
    pub wind_chill_c: Option<f64>,
    pub heat_index_f: Option<f64>,
    pub heat_index_c: Option<f64>,
}

/// Magnus-approximation dewpoint in the unit family of `unit` ("c"/"°C"/
/// "celsius"/"degC"... mean Celsius; anything else Fahrenheit).
pub fn calculate_dewpoint(temperature: Option<f64>, humidity: Option<f64>, unit: &str) -> Option<f64> {
    let (temperature, humidity) = (temperature?, humidity?);
    if humidity <= 0.0 {
        return None;
    }
    let celsius = matches!(
        unit.trim().to_lowercase().as_str(),
        "c" | "celsius" | "°c" | "degc" | "wmounit:degc"
    );
    let ratio = humidity.clamp(0.1, 100.0) / 100.0;
    let temp_c = if celsius {
        temperature
    } else {
        (temperature - 32.0) * 5.0 / 9.0
    };
    let (a, b) = (17.27, 237.7);
    let alpha = (a * temp_c) / (b + temp_c) + ratio.ln();
    let dewpoint_c = (b * alpha) / (a - alpha);
    Some(if celsius {
        dewpoint_c
    } else {
        (dewpoint_c * 9.0 / 5.0) + 32.0
    })
}

/// Humidity rounded to a 0-100 percentage (`fraction` scales 0-1 values).
pub fn normalize_humidity_percent(value: Option<f64>, fraction: bool) -> Option<i64> {
    let numeric = value?;
    let numeric = if fraction { numeric * 100.0 } else { numeric };
    Some(py::round(numeric) as i64)
}

pub fn normalize_temperature_pair(value: Option<f64>, unit: Option<&str>) -> TemperaturePair {
    let (fahrenheit, celsius) = normalize_temperature(value, unit);
    TemperaturePair { fahrenheit, celsius }
}

/// Dewpoint pair, calculated from temperature and humidity when the provider has none.
pub fn normalize_dewpoint_pair(
    dewpoint_value: Option<f64>,
    dewpoint_unit: Option<&str>,
    fallback_temperature_f: Option<f64>,
    humidity_percent: Option<f64>,
) -> TemperaturePair {
    let dewpoint = normalize_temperature_pair(dewpoint_value, dewpoint_unit);
    if dewpoint.fahrenheit.is_some() || dewpoint.celsius.is_some() {
        return dewpoint;
    }
    let dewpoint_f = calculate_dewpoint(fallback_temperature_f, humidity_percent, "fahrenheit");
    TemperaturePair {
        fahrenheit: dewpoint_f,
        celsius: convert_f_to_c(dewpoint_f),
    }
}

pub fn normalize_pressure_pair(value: Option<f64>, unit: Option<&str>) -> PressurePair {
    let (inches, millibars) = normalize_pressure(value, unit);
    PressurePair { inches, millibars }
}

/// Pressure in Pascals for NWS-compatible mapped payloads.
pub fn normalize_pressure_to_pascals(value: Option<f64>, unit: Option<&str>) -> Option<f64> {
    let numeric = value?;
    let unit = unit.unwrap_or("").trim().to_lowercase();
    Some(if unit.contains("hpa") || unit.contains("mb") {
        numeric * 100.0
    } else if unit.contains("pa") {
        numeric
    } else if unit.contains("inch") || unit == "in" || unit == "inhg" {
        numeric * 3386.39
    } else {
        numeric * 100.0
    })
}

/// Values already in millibars/hectopascals.
pub fn normalize_millibars(value: Option<f64>) -> PressurePair {
    PressurePair {
        inches: value.map(|v| v / MB_PER_INHG),
        millibars: value,
    }
}

fn canonical_wind_unit(unit: Option<&str>) -> Option<&str> {
    let text = unit.unwrap_or("").trim().to_lowercase();
    match text.as_str() {
        "m/s" | "meter/s" | "meters/s" | "metre/s" | "metres/s" => Some("wmoUnit:m_s-1"),
        "km/h" | "kmh" | "kph" => Some("wmoUnit:km_h-1"),
        "mph" | "mi/h" => Some("wmoUnit:mi_h-1"),
        "kn" | "kt" | "knot" | "knots" => Some("wmoUnit:kn"),
        _ => unit,
    }
}

pub fn normalize_speed_pair(value: Option<f64>, unit: Option<&str>) -> SpeedPair {
    if value.is_none() {
        return SpeedPair::default();
    }
    let (mph, kph) = convert_wind_speed_to_mph_and_kph(value, canonical_wind_unit(unit));
    SpeedPair { mph, kph }
}

/// Provider-native speed for text forecast fields, e.g. "10 mph".
pub fn format_speed(value: Option<f64>, unit_label: &str) -> Option<String> {
    Some(format!("{} {unit_label}", py::round(value?) as i64))
}

pub fn normalize_visibility_pair(
    value: Option<f64>,
    unit: Option<&str>,
    cap_miles: Option<f64>,
) -> VisibilityPair {
    let Some(numeric) = value else {
        return VisibilityPair::default();
    };
    let unit = unit.unwrap_or("").trim().to_lowercase().replace(' ', "_");
    let (mut miles, mut kilometers) = if unit.contains("ft") || unit.contains("feet") {
        let miles = numeric / 5280.0;
        (miles, miles * KM_PER_MILE)
    } else if unit.contains("km") || unit.contains("kilometer") || unit.contains("kilometre") {
        (numeric / KM_PER_MILE, numeric)
    } else if matches!(unit.as_str(), "mi" | "mile" | "miles") {
        (numeric, numeric * KM_PER_MILE)
    } else {
        (numeric / 1609.344, numeric / 1000.0)
    };
    if let Some(cap) = cap_miles {
        miles = miles.min(cap);
        kilometers = miles * KM_PER_MILE;
    }
    VisibilityPair {
        miles: Some(miles),
        kilometers: Some(kilometers),
    }
}

/// Split a feels-like value into wind chill (colder) or heat index (warmer).
pub fn classify_apparent_temperature(
    temperature_f: Option<f64>,
    apparent_f: Option<f64>,
    apparent_c: Option<f64>,
) -> ApparentTemperatureClassification {
    let (Some(apparent), Some(temperature)) = (apparent_f, temperature_f) else {
        return ApparentTemperatureClassification::default();
    };
    if apparent < temperature {
        ApparentTemperatureClassification {
            wind_chill_f: apparent_f,
            wind_chill_c: apparent_c,
            ..Default::default()
        }
    } else if apparent > temperature {
        ApparentTemperatureClassification {
            heat_index_f: apparent_f,
            heat_index_c: apparent_c,
            ..Default::default()
        }
    } else {
        ApparentTemperatureClassification::default()
    }
}

pub fn pirate_temperature_unit(units: &str) -> &'static str {
    if units == "us" {
        "F"
    } else {
        "C"
    }
}

pub fn pirate_wind_unit(units: &str) -> &'static str {
    match units {
        "us" | "uk2" => "mph",
        "ca" => "km/h",
        _ => "m/s",
    }
}

pub fn pirate_visibility_unit(units: &str) -> &'static str {
    if matches!(units, "us" | "uk2") {
        "mi"
    } else {
        "km"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: Option<f64>, b: f64) -> bool {
        a.is_some_and(|a| (a - b).abs() < 0.01)
    }

    #[test]
    fn temperature_pair_preserves_both_units() {
        let p = normalize_temperature_pair(Some(20.0), Some("°C"));
        assert_eq!(p.fahrenheit, Some(68.0));
        assert_eq!(p.celsius, Some(20.0));
    }

    #[test]
    fn humidity_supports_fraction_and_percent() {
        assert_eq!(normalize_humidity_percent(Some(0.655), true), Some(66));
        assert_eq!(normalize_humidity_percent(Some(72.5), false), Some(72));
        assert_eq!(normalize_humidity_percent(None, false), None);
    }

    #[test]
    fn dewpoint_prefers_provider_value() {
        let p = normalize_dewpoint_pair(Some(10.0), Some("C"), Some(70.0), Some(50.0));
        assert_eq!(p.celsius, Some(10.0));
        let p = normalize_dewpoint_pair(None, Some("F"), Some(70.0), Some(50.0));
        assert!(close(p.fahrenheit, 50.5));
        let p = normalize_dewpoint_pair(None, Some("F"), Some(70.0), Some(0.0));
        assert_eq!(p, TemperaturePair::default());
    }

    #[test]
    fn pressure_pairs_and_pascals() {
        assert_eq!(normalize_pressure_to_pascals(Some(1013.0), Some("hPa")), Some(101300.0));
        assert_eq!(normalize_pressure_to_pascals(Some(101300.0), Some("Pa")), Some(101300.0));
        assert!(close(
            normalize_pressure_to_pascals(Some(29.92), Some("inHg")),
            101320.79
        ));
        assert!(close(normalize_millibars(Some(1013.0)).inches, 29.91));
    }

    #[test]
    fn speed_pair_accepts_provider_labels() {
        let p = normalize_speed_pair(Some(10.0), Some("km/h"));
        assert!(close(p.mph, 6.21));
        assert_eq!(p.kph, Some(10.0));
        let p = normalize_speed_pair(Some(10.0), Some("m/s"));
        assert!(close(p.mph, 22.37));
        assert_eq!(format_speed(Some(10.5), "mph").as_deref(), Some("10 mph"));
        assert_eq!(format_speed(Some(11.5), "mph").as_deref(), Some("12 mph"));
    }

    #[test]
    fn visibility_capped_and_uncapped() {
        let p = normalize_visibility_pair(Some(80000.0), Some("ft"), Some(10.0));
        assert_eq!(p.miles, Some(10.0));
        let p = normalize_visibility_pair(Some(16.09344), Some("km"), None);
        assert!(close(p.miles, 10.0));
        let p = normalize_visibility_pair(Some(24140.16), None, None);
        assert!(close(p.miles, 15.0));
    }

    #[test]
    fn apparent_temperature_classification() {
        let c = classify_apparent_temperature(Some(30.0), Some(20.0), Some(-6.7));
        assert_eq!(c.wind_chill_f, Some(20.0));
        let c = classify_apparent_temperature(Some(90.0), Some(100.0), None);
        assert_eq!(c.heat_index_f, Some(100.0));
        let c = classify_apparent_temperature(Some(50.0), Some(50.0), None);
        assert_eq!(c, ApparentTemperatureClassification::default());
    }

    #[test]
    fn pirate_unit_groups() {
        assert_eq!(pirate_temperature_unit("us"), "F");
        assert_eq!(pirate_temperature_unit("si"), "C");
        assert_eq!(pirate_wind_unit("uk2"), "mph");
        assert_eq!(pirate_wind_unit("ca"), "km/h");
        assert_eq!(pirate_wind_unit("si"), "m/s");
        assert_eq!(pirate_visibility_unit("us"), "mi");
        assert_eq!(pirate_visibility_unit("ca"), "km");
    }
}
